### PR TITLE
hyperscan: support aarch64 platform

### PR DESCRIPTION
**Pull request by tqltech**
Add aarch64 platform branch, which is fully compatible with armv8-a, and at the same time ensure that the use of x86 platform is not affected.

The modification can be divided into the following aspects :
(1) CMakeLists.txt: add branches to judge x86 and aaarch64 platforms.
(2) simd_arm.h: Use the neon instruction to implement 128 operation functions.
(3) simd_x86.h: Save the original simd_utils.h file contents.
(4) simd_utils.h: Select different header files (simd_x86.h or simd_arm.h) according to the platform judgment results.
(5) Intrinsic function: There are multiple calls to the intrinsic function in the code, and the intrinsic function is encapsulated in the modification.
(6) control_verbs.cpp / ExpressionParser.cpp / control_verbs.cpp / Parser.cpp: There are errors in generating these four files under the aarch64 platform. So the files are generated on the x86 platform and copied to the corresponding directory for use. Does not affect the functionality of the x86 platform.

### CODE REVIEW

**Comment by bzhaoopenstack** [comment]
Let's wait for the feedback from upstream side. I'm trying to create a connection with them. If upstream team could review and give some reviews, I'm afraid that the code should be changed. So this PR IMO, leave it here for now.

File: CMakeLists.txt [hunk 1]
```diff
@@ -130,8 +131,8 @@ CMAKE_DEPENDENT_OPTION(DUMP_SUPPORT "Dump code support; normally on, except in r
 
 CMAKE_DEPENDENT_OPTION(DISABLE_ASSERTS "Disable assert(); Asserts are enabled in debug builds, disabled in release builds" OFF "NOT RELEASE_BUILD" ON)
 
-option(BUILD_AVX512 "Experimental: support avx512 in the fat runtime"
-    OFF)
+option(BUILD_AVX512 "Experimental: support avx512 in the fat runtime"
```
Comment (Yikun) [inline CMakeLists.txt:134]: Could we cleanup unrealated changes?

Comment (bzhaoopenstack) [inline CMakeLists.txt:134]: Mark for unecessary change

File: CMakeLists.txt [hunk 2]
```diff
@@ -317,7 +347,10 @@ if (CMAKE_SYSTEM_NAME MATCHES "Linux")
             (CMAKE_VERSION VERSION_GREATER "3.0" AND CMAKE_GENERATOR MATCHES "Ninja")))
         message (STATUS "Building the fat runtime requires the Unix Makefiles generator, or Ninja with CMake v3.0 or higher")
         set (FAT_RUNTIME_REQUISITES FALSE)
-    else()
+    elseif (CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64")
+        message(STATUS "AARCH64 platform don't support fat runtime")
+        set (FAT_RUNTIME_REQUISITES FALSE)
+	else()
```
Comment (bzhaoopenstack) [inline CMakeLists.txt:353]: The indentation of this line is not aligned with the L350

File: CMakeLists.txt [hunk 3]
```diff
@@ -346,6 +381,7 @@ set(C_FLAGS_TO_CHECK
  "-Wstrict-prototypes"
  "-Wmissing-prototypes"
 )
+
```
Comment (bzhaoopenstack) [inline CMakeLists.txt:384]: Mark for unecessary change

File: CMakeLists.txt [hunk 4]
```diff
@@ -469,6 +507,7 @@ if (CORRECT_PCRE_VERSION AND PCRE_BUILD_SOURCE AND BUILD_STATIC_LIBS)
 endif()
 
 add_subdirectory(unit)
+
```
Comment (bzhaoopenstack) [inline CMakeLists.txt:510]: Mark for unecessary change

File: tools/hscollider/ColliderCorporaParser.cpp [hunk 5]
```diff
@@ -0,0 +1,474 @@
+
+
```
Comment (bzhaoopenstack) [inline tools/hscollider/ColliderCorporaParser.cpp:2]: code begins from top of a file

File: unit/internal/simd_utils.cpp [hunk 6]
```diff
@@ -40,7 +40,7 @@ using namespace ue2;
 namespace {
 
 // Switch one bit on in a bitmask.
-template<class Mask>
+template <class Mask>
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:43]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 7]
```diff
@@ -148,7 +148,7 @@ m256 simd_lshift64(const m256 &a, unsigned i) { return lshift64_m256(a, i); }
 m384 simd_lshift64(const m384 &a, unsigned i) { return lshift64_m384(a, i); }
 m512 simd_lshift64(const m512 &a, unsigned i) { return lshift64_m512(a, i); }
 
-template<typename T>
+template <typename T>
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:151]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 8]
```diff
@@ -260,9 +260,9 @@ TYPED_TEST(SimdUtilsTest, or2) {
 
     for (unsigned j = 0; j < 8; j++) {
         for (unsigned i = 0; i < 32; i++) {
-            m256 x = setbit<m256>(j*32+i);
+            m256 x = setbit<m256>(j * 32 + i);
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:263]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 9]
```diff
@@ -260,9 +260,9 @@ TYPED_TEST(SimdUtilsTest, or2) {
...
             m256 y = zeroes256();
-            ASSERT_EQ(1U << j, diffrich256(x, y)) << "bit " << j*32+i << " not happy";
+            ASSERT_EQ(1U << j, diffrich256(x, y)) << "bit " << j * 32 + i << " not happy";
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:265]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 10]
```diff
@@ -431,8 +431,8 @@ TYPED_TEST(SimdUtilsTest, testbit) {
     for (unsigned int i = 0; i < total_bits; i++) {
         TypeParam a = setbit<TypeParam>(i);
         for (unsigned int j = 0; j < total_bits; j++) {
-            ASSERT_EQ(i == j ? 1 : 0, simd_testbit(a, j)) << "bit " << i
-                                                          << " is wrong";
+            ASSERT_EQ(i == j ? 1 : 0, simd_testbit(a, j))
+                << "bit " << i << " is wrong";
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:435]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 11]
```diff
@@ -455,7 +455,6 @@ TYPED_TEST(SimdUtilsTest, setbit) {
         simd_setbit(&a, i);
     }
     ASSERT_FALSE(simd_diff(simd_ones(), a));
-
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:458]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 12]
```diff
@@ -663,12 +662,11 @@ TEST(SimdUtilsTest, movq) {
     ASSERT_EQ(0, memcmp(cmp, &simd, sizeof(simd)));
     ASSERT_EQ(0, memcmp(cmp, &r, sizeof(r)));
 
-    simd = _mm_set_epi64x(~0LL, 0x123456789abcdef);
+    simd = set64x2(~0LL, 0x123456789abcdef);
     r = movq(simd);
     ASSERT_EQ(r, 0x123456789abcdef);
 }
 
-
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:671]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 13]
```diff
@@ -680,7 +678,7 @@ TEST(SimdUtilsTest, set16x8) {
 }
 
 TEST(SimdUtilsTest, set4x32) {
-    u32 cmp[4] = { 0x12345678, 0x12345678, 0x12345678, 0x12345678 };
+    u32 cmp[4] = {0x12345678, 0x12345678, 0x12345678, 0x12345678};
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:681]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 14]
```diff
@@ -714,51 +712,51 @@ TEST(SimdUtilsTest, variableByteShift128) {
...
+    EXPECT_TRUE(
+        !diff128(lshiftbyte_m128(in, 6), variable_byte_shift_m128(in, 6)));
+    EXPECT_TRUE(
+        !diff128(lshiftbyte_m128(in, 7), variable_byte_shift_m128(in, 7)));
+    EXPECT_TRUE(
+        !diff128(lshiftbyte_m128(in, 8), variable_byte_shift_m128(in, 8)));
+    EXPECT_TRUE(
+        !diff128(lshiftbyte_m128(in, 9), variable_byte_shift_m128(in, 9)));
+    EXPECT_TRUE(
+        !diff128(lshiftbyte_m128(in, 10), variable_byte_shift_m128(in, 10)));
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:759]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 15]
```diff
@@ -785,12 +783,12 @@ TEST(SimdUtilsTest, min_u8_m128) {
...
 
 TEST(SimdUtilsTest, sadd_u8_m128) {
-    unsigned char base1[] = {0, 0x80, 0xff, 'A', '1', '2', '3', '4',
-                             '1', '2', '3', '4', '1', '2', '3', '4'};
-    unsigned char base2[] = {'a', 0x80, 'b', 'A', 0x10, 0x10, 0x10, 0x10,
-                             0x30, 0x30, 0x30, 0x30, 0, 0, 0, 0};
+    unsigned char base1[] = {0,   0x80, 0xff, 'A', '1', '2', '3', '4',
+                             '1', '2',  '3',  '4', '1', '2', '3', '4'};
+    unsigned char base2[] = {'a',  0x80, 'b',  'A',  0x10, 0x10, 0x10, 0x10,
+                             0x30, 0x30, 0x30, 0x30, 0,    0,    0,    0};
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:789]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 16]
```diff
@@ -785,12 +783,12 @@ TEST(SimdUtilsTest, min_u8_m128) {
...
     unsigned char expec[] = {'a', 0xff, 0xff, 0x82, 'A', 'B', 'C', 'D',
-                             'a', 'b', 'c', 'd', '1', '2', '3', '4'};
+                             'a', 'b',  'c',  'd',  '1', '2', '3', '4'};
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:791]: Mark for unecessary change

File: unit/internal/simd_utils.cpp [hunk 17]
```diff
@@ -799,11 +797,11 @@ TEST(SimdUtilsTest, sadd_u8_m128) {
...
-                             'a', 'b', 'c', 'd', '1', '2', '3', '4'};
-    unsigned char base2[] = {0, 0x80, 0xff, 'A', '1', '2', '3', '4',
-                             '1', '2', '3', '4', '1', '2', '3', '4'};
-    unsigned char expec[] = {'a', 0x7f, 0, 'A', 0x10, 0x10, 0x10, 0x10,
-                             0x30, 0x30, 0x30, 0x30, 0, 0, 0, 0};
+                             'a', 'b',  'c',  'd',  '1', '2', '3', '4'};
+    unsigned char base2[] = {0,   0x80, 0xff, 'A', '1', '2', '3', '4',
+                             '1', '2',  '3',  '4', '1', '2', '3', '4'};
+    unsigned char expec[] = {'a',  0x7f, 0,    'A',  0x10, 0x10, 0x10, 0x10,
+                             0x30, 0x30, 0x30, 0x30, 0,    0,    0,    0};
```
Comment (bzhaoopenstack) [inline unit/internal/simd_utils.cpp:804]: Mark for unecessary change

File: util/ExpressionParser.cpp [hunk 18]
```diff
@@ -0,0 +1,397 @@
+
+
```
Comment (bzhaoopenstack) [inline util/ExpressionParser.cpp:2]: code should begin from the top of a file

File: CMakeLists.txt [hunk 19]
```diff
@@ -181,26 +182,33 @@ else()
         # generic, which isn't very good in some cases. march=native looks at
         # cpuid info and then chooses the best microarch it can (and replaces
         # the flag), so use that for tune.
+        
+        if (CMAKE_SYSTEM_PROCESSOR MATCHES "x86_64|i386")
```
Comment (Yikun) [inline CMakeLists.txt:186]: I just wanna sure does it include all PROCESSOR flags which we want to cover? i386|i686|amd64|x86_64|AMD64

See ref in [1]

[1] https://github.com/ceph/ceph/blob/master/cmake/modules/SIMDExt.cmake#L48

File: CMakeLists.txt [hunk 20]
```diff
@@ -181,26 +182,33 @@ else()
...
-        execute_process(COMMAND ${CMAKE_C_COMPILER} ${EXEC_ARGS}
-            OUTPUT_QUIET ERROR_QUIET
-            INPUT_FILE /dev/null
-            RESULT_VARIABLE GNUCC_TUNE_TEST)
-        if (NOT GNUCC_TUNE_TEST EQUAL 0)
-            message(SEND_ERROR "Something went wrong determining gcc tune: -mtune=${GNUCC_ARCH} not valid")
+        if (CMAKE_SYSTEM_PROCESSOR MATCHES "aarch64|AARCH64") 
+            set (EXEC_ARGS ${CC_ARG1} -c -Q --help=target -march=armv8-a -mtune=armv8-a)
         endif()
+
```
Comment (Yikun) [inline CMakeLists.txt:211]: unrelated new line

**Review by bzhaoopenstack — COMMENTED** [review]
These reviews could be hold here, just for marking, these won't stop this patch to be merged.

**Review by Yikun — CHANGES_REQUESTED** [review]

**Pull request closed without merging**